keps/sig-scheduling/3521-pod-schedule-readiness/README.md

```diff
@@ -0,0 +1,1032 @@
+or not. This enable scheduler plugin developers to tryout the feature even in Alpha, by crafting
+different enqueue plugins and wire with custom fields or conditions.
+
+- **New column in kubectl:** As we don't write unschedulable condition to a Pod's status if it carries
+`.spec.schedulingPaused=true`, to provider better UX, we're going to add a column "schedule paused"
+to the output of `kubectl get pod` to indicate whether it's scheduling paused or not.
+    > Mentioned in "v1 guarantee" section, if you notice some scheduling(running) Pod's "schedule paused"
+    column is true, you may need to report to your administrator.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+- Scheduler doesn't actively clear a Pod's `schedulingPaused` state. This means if the controller
+logic is ill-implemented, some Pods may stay in Pending state incorrectly. If you noticed a Pod stays
+in Pending state for a long time and it carries `schedulingPaused=true`, you may find out which component
+updates the state (via `.spec.managedFields`) and report the symptom to the component owner.
+
+- Faulty controllers may forget to unset the Pod's `schedulingPaused` state, and hence results in
+a large number of unschedulable Pods. In Alpha, we don't limit the number of unschedulabe Pods caused
+by potential faulty controllers. We will evaluate necessary options in the future to mitigate
+potential abuse.
+
+- End-users may be confused by no scheduling events in the output of `kubectl describe pod xyz`. We
+will provide detailed documentation, along with metrics, tooling (kubectl) and(or) events.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### API
+
+A new API field `schedulingPaused` will be added to Pod's spec:
+
+```go
+type PodSpec struct {
+    <<[UNRESOLVED]>>
+    // Each scheduling gate corresponding with the condition name of .status.conditions.
+    // Gating condition can be updated by different controller, and scheduling is triggered
+    // only when all gating conditions are ready/true.
+    // This option will need to be accompanied with a yet-to-be-implemented fined-grained
+    // permission (https://docs.google.com/document/d/11g9nnoRFcOoeNJDUGAWjlKthowEVM3YGrJA3gLzhpf4)
+    // and consistent with how finalizes/liens work today.
+    SchedulingGates []string
```

Comment:
@smarterclayton it seems that now you are leaning towards this option vs a flag, correct? If so, we also need to discuss the format, we now have two:

1) similar to finalizers: actors add to the list and later remove from the same list
2) similar to readinessGates: actors add to the list in the spec, and later add to a parallel list in status

Regarding option 2, do we have concerns that it will require controllers to have update permissions on pod status? I guess not if we allow updating the list only through a dedicated endpoint since we want to enable fine grained permissions.

I like option 2 because it preserves the history of who decided that a pod is ready, which is useful for debugging and tracking feature usage.

Reply:
Personally, I'm in favor of `readinessGates` as (a) we have pod readiness gates as precedence, and (b) each controller don't need to worry about updating the same API field, and thus keep some factors like ordering away. However, two aspects are still unresolved (https://github.com/kubernetes/enhancements/pull/3522#discussion_r976723138)

- Permission. It's still unclear to me. The [pod readiness gates](https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/580-pod-readiness-gates/README.md) KEP didn't mention about permission. So I really hope @smarterclayton and @liggitt can shed some light - as if we think permission is mandatory here, we have to make it right since day 1.
- Restrictive condition transition. I suppose it's doable to enforce the condition that is transited from NotReady to Ready only. But reposted here for confirmation.

Reply:
I need to read the arguments around pod readiness gates and the whole status thing.  Scheduling happens before ownership is transitioned to a node, I would probably argue that like finalizers, all mutations happen within spec (i view metadata as generic spec in this scenario).

Anyone who can set spec.nodeName is already as powerful as someone who can set the proposed scheduling gates.  OpenShift and some others had admission checks that limited who could mutate that field to effectively cluster admins (roughly the [mutation of that field required an RBAC check](https://github.com/openshift/kubernetes/blob/master/openshift-kube-apiserver/admission/scheduler/podnodeconstraints/admission.go#L135) you have to have the binding subresource permission of a scheduler).

Subresource permission is supported today, so adding a subresource is the easiest way, but we'd probably have to have special cases for what characteristic would be permission (i.e. a check on `readiness/<name>`), but it's not terribly difficult today to make the sub resource `pods/readiness/<name>`.  The more we start bulk updating gates (for instance if the scheduler implements a set of gates via plugins and we want to bulk submit, vs letting individual actors coordinate) the more we get closer to just performing field level checks in an admission handler.  

How would we model multiple gates coordinating in a single scheduler?  I would very much dislike having multiple individual writes to the object, but anything that happens in a single process and is fairly quick should probably just be a single gate (even if it's implemented as a composite of steps). 

Ultimately subresources work when the use case is narrow and does not composite across fields - this is starting to cross that line for binding if we expect lots of gates from a single client being set in a row.

Reply:
> The more we start bulk updating gates (for instance if the scheduler implements a set of gates via plugins and we want to bulk submit, vs letting individual actors coordinate) the more we get closer to just performing field level checks in an admission handler.

I don't think the scheduler itself will have plugins that update gates, the scheduler will only respect those gates (i.e., if they are present, it will simply not attempt to schedule the pod). 

The gates will be updated by external controllers, like a queueing controller or a statefulset migration controller, those by design will likely be controllers that don't share memory and run as separate processes. I think the proper design pattern should be that a controller only handles/updates a single gate.

Reply:
No matter if the field is a single boolean or a list of strings, a Pod's creation is the _only_ chance to set the **complete** gating condition(s). For a single boolean, `true` represents the condition; while for a list of strings, it's highly customized list. After a Pod's creation, only "semantical subtraction" can be allowed: for boolean, it's true -> false; for a list of strings, it's "delete a particular element from the list" (it just behaves differently for finalizer-like list vs. readiness gates-like list)

> I would very much dislike having multiple individual writes to the object

For the design of a list of strings, both options (finalizer-like and readiness gate-like) will inevictably yield N individual writes to a Pod - each controller needs to spawn an API call to relax the readiness gate anyway (either by deleting one entry from the finalizer-like lists, or updating their particular .status.conditions[x] to ready/true).

Reply:
Sorry, I should have said "N writes by the same client".  "N writes by N clients" is fine.  "2N writes by N clients"... might be ok, but I want to ask here.  What sort of gates would we imagine?  Someone filling out scheduling constraints (fleshing out topologyConstraints)?  Adding tolerations?  Adding resources?  The more general purpose, the more I want to at least consider avoiding `2N` writes in the design of the mechanism.  The more general purpose, the more we want to avoid `MN` RBAC checks on fields.  I like general purpose mutation control, but I hate to overly couple a new design like this with it, as we noted in the subresource doc.

Just to be 100% certain, we don't think we'd have an internal to the scheduler gate?

Reply:
> What sort of gates would we imagine?

The gates can be related to in-house multi-tenancy (e.g., advanced quota control), batch support (e.g., gang-scheduling so start scheduling only when a quorum is satisfied), etc.

> I want to at least consider avoiding 2N writes in the design of the mechanism

I'm actually more aggressive to restrict it a N writes at maximum. That's to say, each gate owner has only one chance to turn the flag from paused to unpaused.

> Just to be 100% certain, we don't think we'd have an internal to the scheduler gate?

Do you mean if scheduler would be manipulating the gates? No, scheduler would not mutate a Pod's spec in any way. The only fields it update is status.condition, status.nominatedNodeName and v1.binding (.spec.nodeName).

>> a Pod's creation is the only chance to set the complete gating condition(s). 
>
> I didn't anticipate this being required. Gate 1 could add 2 more gates and remove itself. That was something we believed we should support in initializers, because it allowed chaining at the cost of more stages. Is there a reason not to allow it?

That makes the scheduling racy. Suppose all gates are lifted, and the scheduler to schedule a Pod, and meanwhile then new gate(s) comes in aiming to pause the Pod, but due to current scheduling design, that "gating" signal cannot stop an in-flight pod's scheduling. 

This is also consistent what [Pod Readiness Gate](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/580-pod-readiness-gates#implementation-detailsnotesconstraints): "ReadinessGates can only be specified at pod creation."

> Do we have a use case that requires gates to be processed in order, vs being parallel?

For now I'm not aware of any req on this. Say, we assume each gate's logic is independent.

Reply:
> That makes the scheduling racy. Suppose all gates are lifted, and the scheduler to schedule a Pod, and meanwhile then new gate(s) comes in aiming to pause the Pod, but due to current scheduling design, that "gating" signal cannot stop an in-flight pod's scheduling.

Sorry, I meant "the success of gate 1 is indicated at the same time gate 2 and 3 are added".  That is not racy because the scheduler observes version 1 with "gate1", then version 2 with "gate1 (complete), gate 2, gate 3".  But you're right, for us to deviate from readiness gates consistency we'd need a use case justifying it.

Even though I don't like mutating webhooks injecting things on pod creation (every additional webhook reduces total cluster reliability because we've added a new independent source of failure), an approach where a third party used mutable gates to behave like initializers (instead of doing it at creation time, you do it via a regular controller) changes it from a problem of the api client (get a fail closed on the webhook) to a problem visible to the scheduler (the gate initializer controller hangs).  That said, if we're going to have CEL eventually, you could use CEL to set up your basic gates and let a more sophisticated gate implementation (the batch scheduler) handle the details, including potential delegation.

I guess the question is - if we disallow addition the gates post-creation now, can we later on relax it safely?  I think so, in which case we can start with the more limited option and see how we go.

Reply:
I agree that we don't need to enforce order, the base use case is a group of independent controllers that don't really coordinate.

> I guess the question is - if we disallow addition the gates post-creation now, can we later on relax it safely? I think so, in which case we can start with the more limited option and see how we go.

+1, I think it is safe to start strict and potentially relax later, especially if this field will be gated via RBAC.

At this point I think we have two issues to resolve:
1. API format: finalizer vs readinessGate format. How can we move forward from here? 
2. How to gate this field, subresource can be mentioned as an option for now, but we could also say that we will wait for a recommendation from sig-auth.

Reply:
> "the success of gate 1 is indicated at the same time gate 2 and 3 are added"

If the 3 operations are completed within one API call, then yes it's not racy, but as you mentioned, we may start with the most strict version as a start.

> I guess the question is - if we disallow addition the gates post-creation now, can we later on relax it safely? I think so, in which case we can start with the more limited option and see how we go.

+1 as well.

> 1. API format: finalizer vs readinessGate format. How can we move forward from here?

For readinessGates, to confirm we're on the same page that:

- a) each gate corresponds to a condition with the same name as the condition's `type`
- b) specifying gates as Pod's creation time indicates the corresponding conditions are "paused" initially
- c) nil conditions{gates[x]} is treated as a legal "paused" condition
- d) there will be validation logic enforcing conditions{gates[x]} is updated in an one-way ("paused"->"unpaused") direction
- e) in Alpha, `gates` are immutable after creation time

If we all agree on this, I'd be glad to update the doc to turn a single boolean to this.

> 2\. How to gate this field, subresource can be mentioned as an option for now, but we could also say that we will wait for a recommendation from sig-auth.

A practical way is to document the auth restriction in Alpha, and revisit it in Beta (hope we will have a canonical auth way by then). WDYT?

Reply:
Nope, we don't expect individual end-user to patch the status. Only privileged/trusted controllers can patch.

Reply:
Yes, correct. Regarding Pods that are not subject to schedulingGates, they are scheduled as before.

Reply:
One potential issue is if by mistake the same gate name was used for both pod and scheduling readiness. Perhaps we should validate that readinessGates and schedulingReadinessGates are mutually exclusive.

Reply:
yes, currently every readiness gate needs to be qualified, so no harm to add additional validation checks on the uniqueness of gate names.

Reply:
@smarterclayton as well. We may tend to adopt schedulingGates over finalizer-like string list.

Reply:
Re: string list vs array list - it is strictly preferable in almost all cases on a single type to bias away from `[]primitive` to `[]struct` - there is almost no downside because the field can be extensible and the serialization isn't substanitally more complex.  

For finalizers, the argument came down to whether a field in metadata would be extensible - we could not envision it (it's really hard to extend finalizer generically) and we had annotations and labels to carry the extensibility for custom finalizers if we needed it (or per type fields).  So I would argue finalizers are special because they're in metadata, while every other type should prefer `[]struct`.

I'm ok with permission bits coming after alpha, it's no worse than finalizers, leins, spec.nodeName, or a few other fields today.

Re: mutually exclusive - just to verify, is the problem that the gate names go into the same list somewhere?  I don't see how that's possible, readiness is a different part of the lifecycle than scheduling (by definition, a pod that hasn't been scheduled can't be ready)?

Reply:
> Re: string list vs array list

The only concern is it's not that consistent with the API design of current readiness gates. Readiness gates are of type []string, and each gate corresponds to status.conditions[x]. But personally I have no strong option on []string over []struct. Given readiness gates have been a while, not quite its API pattern is still a best practice. @liggitt @ahg-g 

> just to verify, is the problem that the gate names go into the same list somewhere?

It's based on using `[]string` as the schedulingGates, so the mapped condition needs to be unique - each gate name (scheduling or regular for network) corresponds to a condition with its `type` as the same name (required to be a qualified naming style: foo.com/xyz).

Reply:
@smarterclayton I don't think you explicitly made your recommendation regarding finalizers vs readinessGates format. The former is a list that we remove from, the latter is a static list of gates and PodConditions are used to "open" the gates.

> Re: mutually exclusive - just to verify, is the problem that the gate names go into the same list somewhere? I don't see how that's possible, readiness is a different part of the lifecycle than scheduling (by definition, a pod that hasn't been scheduled can't be ready)?

Mutually exclusive is relevant if we go with the podReadiness format. The lists where we declare the gates are separate in the pod Spec, but both features will use PodConditions to "open" the gates. Think of the case where the same gate is declared in both lists by mistake, for example two unrelated podReadiness and schedulingReadiness controllers using the same gate name.

Reply:
> The former is a list that we remove from, the latter is a static list of gates and PodConditions are used to "open" the gates.

Sorry, I was implying that pod conditions weren't the right approach (no status, so no conditions), so implication being that the modelling would be a `spec.[]struct` that clients remove from OR simply set a field (like finalizers).  I was emphasizing the spec structure of readinessGates vs the "spec + status coordination around conditions".  Conditions are appropriate for multiple clients to coordinate on once the pod passes to node ownership (spec.nodeName is set), but status is not appropriate for multiple clients to coordinate on prior to the pod being scheduled (IMO, open to arguments).

ReadinessGates can be mutable over time, SchedulingGates are one way.  Whether we remove or flag the spec with some status, we should consider whether we need a record on the pod of the gate being passed.  I would probably say we don't need a record (we didn't need a record for finalizers).

Reply:
Actually I agreed with the interpretation of .status.conditions, and that's why I mentioned a couple of time the concern of "enforcing one-way transition on conditions[x]". Enforcing on conditions doesn't quite have a precedence, although in readiness gates's KEP, it mentioned the expectation:

> To conform with this proposals, workload controllers MUST take pod “Ready” condition as the final signal to proceed during transitions

But I checked the code, there is no enforcement logic.

Reply:
> Conditions are appropriate for multiple clients to coordinate on once the pod passes to node ownership (spec.nodeName is set), but status is not appropriate for multiple clients to coordinate on prior to the pod being scheduled (IMO, open to arguments).

I'm not sure I buy the pre/post-node-assignment distinction on using pod.status / pod.status.conditions

Wanting clarity on the implications of building things on top of conditions that expect/require one-way semantics is fair

Reply:
> Can you describe what code you checked? You mean, readinessGate has no enforcement of one way flip?

I checked the following code paths:

- pkg/apis/core/validation/validation.go#validatePodConditions
- pkg/apis/core/validation/validation.go#validateReadinessGates

Also, I tested it on a kind v1.25 env: https://gist.github.com/Huang-Wei/734b468571d8bd3a8e3211f34c197414

Reply:
> I'm not sure I buy the pre/post-node-assignment distinction on using pod.status / pod.status.conditions

Fair, I think it's more "conditions as parts of a state machine feel wrong" vs "pre / post kubelet ownership".

Conditions don't feel like they are the appropriate field to represent a state machine, especially since most gates would be unlikely to need to convey information to a user.  Whereas a failure to go ready in a pod is a more granular operation that users are likely to debug.

Your one-way point is clearer - nothing stops an integrator from adding a condition to indicate failure to proceed, but the semantics of conditions do not include immutability today.  Whereas we can add such semantics to a new field fairly cleanly.  Hopefully per condition type immutability semantics is not seen as desirable for a consumer today (certainly not practical to add in the near term).

As an aside, i'm now worried about readiness gates not taking into account that node state can be lost without the pod being deleted - a pod that has been ready once may have gates on the node that need to be reset in some pod recovery scenarios.  I hope we don't have gates making such an assumption and being incorrect on failure...

> I checked the following code paths

Yeah, someone adding one way immutability to conditions would be questionable - conditions don't have those semantics today, this seems like tightening validation.

Reply:
It looks like now we have a stronger reason not to pursue readinessGates semantics. Although keeping a record of the gates would be really nice for debug-ability, there are obviously other ways we can achieve that (logging and events at the time of pod creation).

> That might be "no other finalizers before me in the list", that could also be coordination with other objects. We should as part of this design consider some normative / recommended coordination documentation similar to finalizers (since finalizers are a stronger phase transition construct than readiness gates).

I guess the only recommended coordination we can document is that dependency between gate managers should be one way to avoid circular dependencies.

Reply:
Re: "conditions as parts of a state machine feel wrong"

Can we bypass the enforcement on conditions just like readiness gates? and also document it that the integrator should be fully responsible for it. In such cases, we have the freedom to reset the condition in some corner cases like Clayton mentioned.

Reply:
Hrm.  In essence, we have a spectrum:

1. Initializers / finalizers - `[]string`, removal only, additional data stored on object or at worst in meta, extremely unlikely we add supporting data
2. schedulingGates - removal only, additional data stored on pod spec, potentially other data in the future but unlkely, users unlikely to need the status of each gate, integrators likely able to edit spec
3. readinessGates - `[]struct + conditions`, readiness is inherently reactionary and two-way, gates say they are one way but are not (and probably can't safely change conditions), it think it is possible to lose all node state in such a way that gates have to be re-executed (but we don't clearly define that anywhere), integrators SHOULD NOT be able to edit spec, users want to see the status of each gate

I think the argument I'm thinking about is that `nothing requires conditions, therefore why are they a part of the flow`.  Consistency is one argument, but I think the use case is closer to initializers / finalizers, and since we don't have a *need*, I'd argue YAGNI.  If we do get feedback that users want a lot of condition state from gates, we could reassess in beta (the structure wouldn't change for our spec).

Thoughts?

Reply:
I agree with option 2, and I think even now integrators are free to add conditions to the pod if they wish to offer more status information, don't they? So there is no need to tie that to our feature.

Reply:
Right, good argument.

Reply:
Option 2 should work for most cases. If there is some case like re-scheduling a pod without deleting it, it's also extensible / backward-compatible to add fields.

Reply:
updated.